// @effect/platform/dist/cjs/Error.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isPlatformError = exports.TypeIdError = exports.SystemError = exports.PlatformErrorTypeId = exports.BadArgument = void 0;
var Data = _interopRequireWildcard(require("effect/Data"));
var Predicate = _interopRequireWildcard(require("effect/Predicate"));
var internal = _interopRequireWildcard(require("./internal/error.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 1.0.0
 * @category type id
 */
const PlatformErrorTypeId = exports.PlatformErrorTypeId = internal.PlatformErrorTypeId;
/**
 * @since 1.0.0
 * @category refinements
 */
const isPlatformError = u => Predicate.hasProperty(u, PlatformErrorTypeId);
/**
 * @since 1.0.0
 * @category error
 */
exports.isPlatformError = isPlatformError;
const TypeIdError = (typeId, tag) => {
  class Base extends Data.Error {
    _tag = tag;
  }
  ;
  Base.prototype[typeId] = typeId;
  Base.prototype.name = tag;
  return Base;
};
/**
 * @since 1.0.0
 * @category error
 */
exports.TypeIdError = TypeIdError;
const BadArgument = exports.BadArgument = internal.badArgument;
/**
 * @since 1.0.0
 * @category error
 */
const SystemError = exports.SystemError = internal.systemError;
