// @effect/platform-browser/dist/cjs/Permissions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.layer = exports.TypeId = exports.PermissionsError = exports.Permissions = exports.ErrorTypeId = void 0;
var _Error = require("@effect/platform/Error");
var Context = _interopRequireWildcard(require("effect/Context"));
var Effect = _interopRequireWildcard(require("effect/Effect"));
var Layer = _interopRequireWildcard(require("effect/Layer"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category type ids
 */
const TypeId = exports.TypeId = /*#__PURE__*/Symbol.for("@effect/platform-browser/Permissions");
/**
 * @since 1.0.0
 * @category type ids
 */
const ErrorTypeId = exports.ErrorTypeId = /*#__PURE__*/Symbol.for("@effect/platform-browser/Permissions/PermissionsError");
/**
 * @since 1.0.0
 * @category errors
 */
class PermissionsError extends /*#__PURE__*/(0, _Error.TypeIdError)(ErrorTypeId, "PermissionsError") {
  get message() {
    return this.reason;
  }
}
/**
 * @since 1.0.0
 * @category tags
 */
exports.PermissionsError = PermissionsError;
const Permissions = exports.Permissions = /*#__PURE__*/Context.GenericTag("@effect/platform-browser/Permissions");
/**
 * A layer that directly interfaces with the `navigator.permissions` api
 *
 * @since 1.0.0
 * @category layers
 */
const layer = exports.layer = /*#__PURE__*/Layer.succeed(Permissions, /*#__PURE__*/Permissions.of({
  [TypeId]: TypeId,
  query: name => Effect.tryPromise({
    try: () => navigator.permissions.query({
      name
    }),
    catch: cause => new PermissionsError({
      reason: cause instanceof DOMException ? "InvalidStateError" : "TypeError",
      cause
    })
  })
}));
