// @effect/printer-ansi/dist/cjs/Color.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.yellow = exports.white = exports.toCode = exports.red = exports.magenta = exports.green = exports.cyan = exports.blue = exports.black = void 0;
var internal = _interopRequireWildcard(require("./internal/color.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 1.0.0
 */

// -----------------------------------------------------------------------------
// Constructors
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category constructors
 */
const black = exports.black = internal.black;
/**
 * @since 1.0.0
 * @category constructors
 */
const red = exports.red = internal.red;
/**
 * @since 1.0.0
 * @category constructors
 */
const green = exports.green = internal.green;
/**
 * @since 1.0.0
 * @category constructors
 */
const yellow = exports.yellow = internal.yellow;
/**
 * @since 1.0.0
 * @category constructors
 */
const blue = exports.blue = internal.blue;
/**
 * @since 1.0.0
 * @category constructors
 */
const magenta = exports.magenta = internal.magenta;
/**
 * @since 1.0.0
 * @category constructors
 */
const cyan = exports.cyan = internal.cyan;
/**
 * @since 1.0.0
 * @category constructors
 */
const white = exports.white = internal.white;
// -----------------------------------------------------------------------------
// Destructors
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category destructors
 */
const toCode = exports.toCode = internal.toCode;
