// @effect/printer-ansi/dist/cjs/internal/ansiDoc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.eraseUp = exports.eraseStartLine = exports.eraseScreen = exports.eraseLines = exports.eraseLine = exports.eraseEndLine = exports.eraseDown = exports.cursorUp = exports.cursorTo = exports.cursorShow = exports.cursorSavePosition = exports.cursorRestorePosition = exports.cursorPrevLine = exports.cursorNextLine = exports.cursorMove = exports.cursorLeft = exports.cursorHide = exports.cursorForward = exports.cursorDown = exports.cursorBackward = exports.beep = void 0;
var Doc = _interopRequireWildcard(require("@effect/printer/Doc"));
var InternalAnsi = _interopRequireWildcard(require("./ansi.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/** @internal */
const beep = exports.beep = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.beep);
/** @internal */
const cursorTo = (column, row) => Doc.annotate(Doc.empty, InternalAnsi.cursorTo(column, row));
/** @internal */
exports.cursorTo = cursorTo;
const cursorMove = (column, row) => Doc.annotate(Doc.empty, InternalAnsi.cursorMove(column, row));
/** @internal */
exports.cursorMove = cursorMove;
const cursorUp = (lines = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorUp(lines));
/** @internal */
exports.cursorUp = cursorUp;
const cursorDown = (lines = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorDown(lines));
/** @internal */
exports.cursorDown = cursorDown;
const cursorForward = (columns = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorForward(columns));
/** @internal */
exports.cursorForward = cursorForward;
const cursorBackward = (columns = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorBackward(columns));
/** @internal */
exports.cursorBackward = cursorBackward;
const cursorLeft = exports.cursorLeft = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorLeft);
/** @internal */
const cursorSavePosition = exports.cursorSavePosition = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorSavePosition);
/** @internal */
const cursorRestorePosition = exports.cursorRestorePosition = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorRestorePosition);
/** @internal */
const cursorNextLine = (rows = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorNextLine(rows));
/** @internal */
exports.cursorNextLine = cursorNextLine;
const cursorPrevLine = (rows = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorPrevLine(rows));
/** @internal */
exports.cursorPrevLine = cursorPrevLine;
const cursorHide = exports.cursorHide = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorHide);
/** @internal */
const cursorShow = exports.cursorShow = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorShow);
/** @internal */
const eraseLines = rows => Doc.annotate(Doc.empty, InternalAnsi.eraseLines(rows));
/** @internal */
exports.eraseLines = eraseLines;
const eraseEndLine = exports.eraseEndLine = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseEndLine);
/** @internal */
const eraseStartLine = exports.eraseStartLine = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseStartLine);
/** @internal */
const eraseLine = exports.eraseLine = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseLine);
/** @internal */
const eraseDown = exports.eraseDown = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseDown);
/** @internal */
const eraseUp = exports.eraseUp = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseUp);
/** @internal */
const eraseScreen = exports.eraseScreen = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseScreen);
