// @livestore/livestore/src/mod.ts
export { Store } from './store/store.js'
export { createStore, createStorePromise, type CreateStoreOptions } from './store/create-store.js'
export type { QueryDebugInfo, RefreshReason, OtelOptions } from './store/store-types.js'
// We're re-exporting `Schema` from `effect` for convenience
export { Schema } from '@livestore/utils/effect'

export {
  type LiveStoreContext,
  type LiveStoreContextRunning,
  type ShutdownDeferred,
  makeShutdownDeferred,
} from './store/store-types.js'

export { SqliteDbWrapper, emptyDebugInfo } from './SqliteDbWrapper.js'

export {
  queryDb,
  computed,
  signal,
  type LiveQuery,
  type LiveQueryDef,
  type Signal,
  type SignalDef,
  type RcRef,
} from './live-queries/mod.js'

export * from '@livestore/common/schema'
export {
  sql,
  SessionIdSymbol,
  type BootStatus,
  type SqliteDb,
  type DebugInfo,
  type MutableDebugInfo,
  prepareBindValues,
  type Bindable,
  type PreparedBindValues,
  type QueryBuilderAst,
  type QueryBuilder,
  type RowQuery,
  StoreInterrupted,
  IntentionalShutdownCause,
  provideOtel,
} from '@livestore/common'

export { deepEqual } from '@livestore/utils'
export { nanoid } from '@livestore/utils/nanoid'

export * from './utils/stack-info.js'

export type { ClientSession, Adapter, PreparedStatement } from '@livestore/common'
