// @effect/printer/dist/cjs/Flatten.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.neverFlat = exports.map = exports.isNeverFlat = exports.isFlattened = exports.isFlatten = exports.isAlreadyFlat = exports.flattened = exports.alreadyFlat = exports.FlattenTypeId = void 0;
var internal = _interopRequireWildcard(require("./internal/flatten.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category symbol
 */
const FlattenTypeId = exports.FlattenTypeId = internal.FlattenTypeId;
// -----------------------------------------------------------------------------
// Refinements
// -----------------------------------------------------------------------------
/**
 * Returns `true` if the specified value is a `Flatten`, `false` otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isFlatten = exports.isFlatten = internal.isFlatten;
/**
 * Returns `true` if the specified `Flatten` is a `Flattened`, `false` otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isFlattened = exports.isFlattened = internal.isFlattened;
/**
 * Returns `true` if the specified `Flatten` is an `AlreadyFlat`, `false` otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isAlreadyFlat = exports.isAlreadyFlat = internal.isAlreadyFlat;
/**
 * Returns `true` if the specified `Flatten` is a `NeverFlat`, `false` otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isNeverFlat = exports.isNeverFlat = internal.isNeverFlat;
// -----------------------------------------------------------------------------
// Constructors
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category constructors
 */
const flattened = exports.flattened = internal.flattened;
/**
 * @since 1.0.0
 * @category constructors
 */
const alreadyFlat = exports.alreadyFlat = internal.alreadyFlat;
/**
 * @since 1.0.0
 * @category constructors
 */
const neverFlat = exports.neverFlat = internal.neverFlat;
// -----------------------------------------------------------------------------
// Combinators
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category mapping
 */
const map = exports.map = internal.map;
