// @effect/printer/dist/cjs/internal/pageWidth.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unbounded = exports.remainingWidth = exports.isUnbounded = exports.isPageWidth = exports.isAvailablePerLine = exports.defaultPageWidth = exports.availablePerLine = exports.PageWidthTypeId = void 0;
var Equal = _interopRequireWildcard(require("effect/Equal"));
var _Function = require("effect/Function");
var Hash = _interopRequireWildcard(require("effect/Hash"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
// -----------------------------------------------------------------------------
// Models
// -----------------------------------------------------------------------------
const PageWidthSymbolKey = "@effect/printer/PageWidth";
/** @internal */
const PageWidthTypeId = exports.PageWidthTypeId = /*#__PURE__*/Symbol.for(PageWidthSymbolKey);
const protoHash = {
  AvailablePerLine: self => (0, _Function.pipe)(Hash.hash("@effect/printer/PageWidth/AvailablePerLine"), Hash.combine(Hash.hash(PageWidthSymbolKey)), Hash.combine(Hash.hash(self.lineWidth)), Hash.combine(Hash.hash(self.ribbonFraction))),
  Unbounded: _ => (0, _Function.pipe)(Hash.hash("@effect/printer/PageWidth/Unbounded"), Hash.combine(Hash.hash(PageWidthSymbolKey)))
};
const protoEqual = {
  AvailablePerLine: (self, that) => isPageWidth(that) && that._tag === "AvailablePerLine" && self.lineWidth === that.lineWidth && self.ribbonFraction === that.ribbonFraction,
  Unbounded: (self, that) => isPageWidth(that) && that._tag === "Unbounded"
};
const proto = {
  [PageWidthTypeId]: PageWidthTypeId,
  [Hash.symbol]() {
    return Hash.cached(this, protoHash[this._tag](this));
  },
  [Equal.symbol](that) {
    return protoEqual[this._tag](this, that);
  }
};
// -----------------------------------------------------------------------------
// Refinements
// -----------------------------------------------------------------------------
/** @internal */
const isPageWidth = u => typeof u === "object" && u != null && PageWidthTypeId in u;
/** @internal */
exports.isPageWidth = isPageWidth;
const isAvailablePerLine = self => self._tag === "AvailablePerLine";
/** @internal */
exports.isAvailablePerLine = isAvailablePerLine;
const isUnbounded = self => self._tag === "AvailablePerLine";
// -----------------------------------------------------------------------------
// Constructors
// -----------------------------------------------------------------------------
/** @internal */
exports.isUnbounded = isUnbounded;
const availablePerLine = (lineWidth, ribbonFraction) => {
  const op = Object.create(proto);
  op._tag = "AvailablePerLine";
  op.lineWidth = lineWidth;
  op.ribbonFraction = ribbonFraction;
  return op;
};
/** @internal */
exports.availablePerLine = availablePerLine;
const unbounded = exports.unbounded = /*#__PURE__*/(() => {
  const op = /*#__PURE__*/Object.create(proto);
  op._tag = "Unbounded";
  return op;
})();
/** @internal */
const defaultPageWidth = exports.defaultPageWidth = /*#__PURE__*/availablePerLine(80, 1);
// -----------------------------------------------------------------------------
// Utilities
// -----------------------------------------------------------------------------
/** @internal */
const remainingWidth = (pageWidth, ribbonFraction, indentation, currentColumn) => {
  const columnsLeftInLine = pageWidth - currentColumn;
  const ribbonWidth = Math.max(0, Math.min(pageWidth, Math.floor(pageWidth * ribbonFraction)));
  const columnsLeftInRibbon = indentation + ribbonWidth - currentColumn;
  return Math.min(columnsLeftInLine, columnsLeftInRibbon);
};
exports.remainingWidth = remainingWidth;
